test/api/api_test.go

```diff
@@ -138,13 +140,77 @@ func TestWithDBSession(t *testing.T) {
 	}
 }
 
+// TestWithDBSession_RetryOnWriteConflict ensures that WithDBSession can
+// properly retry requests on MongoDB WriteConflict error.
+func TestWithDBSession_RetryOnWriteConflict(t *testing.T) {
```

Comment:
Can you add another, a bit more deterministic test? As a suggestion I think you could create a new dependency which simply forces a handler to return the write conflict error n times.

Then you run the test once with 1 failure, once with `maxRetries` failures and once with `maxRetries+1` failures to make sure it passes twice and fails once.

Reply:
Added `testWithDBSessionRetryOnWriteConflictRepeatedFailures`.

I took a slightly different approach with the retries but I think I captured the intention. If you prefer a test which we actually run three times with different parameters, I can write that too but it seemed overly complicated while achieving the same thing.